### ISSUE TITLE
Improve OpenSeadragon Tile Loading Efficiency for Overlapping Images

### ISSUE THREAD

**Issue by hrghauri** [openseadragon/openseadragon#2338]
When using OpenSeadragon to view multiple images that are stacked on top of each other, it appears that OSD makes network requests for tiles of images that are hidden behind a fully opaque image from the top. This results in unnecessary network traffic and slows down the loading of tiles for the top image. To address this issue, I propose enhancing OSD so that it avoids making requests for tiles that are hidden behind a fully opaque layer from the top, unless the opacity of the top image is less than 1 or if the preload flag is set to true. This would reduce the amount of unnecessary network traffic and improve the loading speed of tiles for overlapping images in OSD.

**Comment by iangilman** [comment]
I agree this would be good! @hrghauri I understand from a conversation in Discord that you might be up for pursuing a fix?

Implementing it will be interesting... Currently a TiledImage determines what tiles to load without consulting any other TiledImages. Adding this functionality will require each TiledImage to check in with the World to see what other images might be overlapping it and where they are. It shouldn't be too hard... You can access it via the TiledImage's `viewer.world`. Images with a higher index than the current image are above that image. 

The mechanism a TiledImage uses to determine if a particular area is already covered by existing tiles (of its own) is its `coverage` property. That may or may not be a good thing to hook into for this... It's already in place, so it would be convenient, but it's structured in terms of tiles and levels, rather than explicit coordinates, so it might not map well. 

At any rate, looking at how `coverage` is dealt with in `TiledImage.js` will at least help you get a sense for the areas this feature will need to touch. 

Right now, the `World` goes through all of the images in index order every time we need to draw a frame. Image 0 is drawn and then image 1, etc. As the images draw, they also determine their coverage and what tile to load next. One possible implementation for this feature would be to break that up into two passes: First, go through the images in reverse order determining coverage and passing the accumulated coverage onto each successive image; second, go through the images in forward order, drawing them.  

One thing to be aware of when considering if an image is fully opaque is whether it has transparency in it. At the moment, we generally just assume images with PNG tiles contain transparency, but of course that's not always true. At any rate, we should honor the `hasTransparency` property. 

What do you think? Let me know if you have questions!

**Comment by hrghauri** [comment]
> I understand from a conversation in Discord that you might be up for pursuing a fix?

Yes, I will be pursuing a fix. However, this will be my first big fix in OSD so bear with me if I have lots of questions. :)

> At any rate, looking at how coverage is dealt with in TiledImage.js will at least help you get a sense for the areas this feature will need to touch.

I did have a quick look at coverage. We will need something similar in the world class, I presume. The world class knows about each item's coordinates.

Also, another question, do you think we would need modify tileSource class for this as well?

> One thing to be aware of when considering if an image is fully opaque is whether it has transparency in it. At the moment, we generally just assume images with PNG tiles contain transparency, but of course that's not always true. At any rate, we should honor the hasTransparency property.

I guess, the condition then becomes that when we have an overlapping area between images, is 'don't make network requests for the bottom image tiles unless opacity of the top layer is below 1 or if the preload flag is set to true or the top layer hasTranparency set to true.

**Comment by iangilman** [comment]
> Yes, I will be pursuing a fix. However, this will be my first big fix in OSD so please bear with me if I have lots of questions. :)

Absolutely, no worries! 

> I did have a quick look at coverage. We will need something similar in the world class, I presume. The world class knows about each item's coordinates.

The World has a list of TiledImages and it can ask each of them for their bounds in viewport coordinates. Also, if we go with the strategy of calculating coverage in a pass before the drawing pass, that information can come out of the calculation function. 

I don't think we would want it structured like the TiledImage's `coverage` which is essentially an array of tiles. Perhaps for this, we would want an array of rectangles in viewport coordinates. The simplest would be a rectangle for the entirety of a TiledImage if it's opaque, and no rectangle if it has any transparency. 

> Also, another question, do you think we would need modify tileSource and imageloader classes for this as well? And any other you can think of?

I don't think those classes would be affected. They just do what the TiledImage asks them to. 

> I guess, the condition then becomes that when we have an overlapping area between images, is 'don't make network requests for the bottom image tiles unless opacity of the top layer is below 1 or if the preload flag is set to true or the top layer hasTranparency set to true.'

This sounds right!
